### PR TITLE
🎉 New Source: Medallia connector

### PR DESCRIPTION
**Pull request by JamesGouldSonos**
## What
This PR proposes a new Medallia connector, the initial iteration provides the ability to sync Fields and Feedback entities.

## How

- The connector is a Python source using the Airbyte CDS
- The only configuration needed is Medallia OAUTH credentials and API endpoint details for your instance.

## 🚨 User Impact 🚨
None
*For connector PRs, use this section to explain which type of semantic versioning bump occurs as a result of the changes. Refer to our [Semantic Versioning for Connectors](https://docs.airbyte.com/contributing-to-airbyte/#semantic-versioning-for-connectors) guidelines for more information. **Breaking changes to connectors must be documented by an Airbyte engineer (PR author, or reviewer for community PRs) by using the [Breaking Change Release Playbook](https://docs.google.com/document/d/1VYQggHbL_PN0dDDu7rCyzBLGRtX-R3cpwXaY8QxEgzw/edit).***

*If there are breaking changes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.*

## Pre-merge Actions
*Expand the relevant checklist and delete the others.*

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- Connector version is set to `0.0.1`
    - `Dockerfile` has version `0.0.1`
- Documentation updated
    - Connector's `README.md`
    - Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - `docs/integrations/<source or destination>/<name>.md` including changelog with an entry for the initial version. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - `docs/integrations/README.md`

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

<details><summary><strong>Connector Generator</strong></summary>

- Issue acceptance criteria met
- PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests#pull-request-title-convention)
- If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- Documentation which references the generator is updated as needed

</details>

### CODE REVIEW

**Comment by marcosmarxm** [comment]
@JamesGouldSonos do you have a test account to share to run tests?

**Comment by JamesGouldSonos** [comment]
Hi Marcos,

Apologies I do have a Medallia account we have been using this with but
this is not a generic account we can share for testing purposes,

KInd Regards,

James

**Comment by marcosmarxm** [comment]
> Connector

@JamesGouldSonos can you share the output from the integration tests to make sure all tests are working?

File: airbyte-integrations/connectors/source-medallia/integration_tests/configured_catalog.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
+{
+  "streams": [
+    {
+      "stream": {
+        "name": "table_name",
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-medallia/integration_tests/configured_catalog.json:5]: This must have the streams you have implemented for the connector.

File: airbyte-integrations/connectors/source-medallia/unit_tests/unit_test.py [hunk 2]
```diff
@@ -0,0 +1,7 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+
+def test_example_method():
+    assert True
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-medallia/unit_tests/unit_test.py:7]: Please add unit test for the custom auth and the schema validation.

File: airbyte-integrations/connectors/source-medallia/source_medallia/schemas/feedback.json [hunk 3]
```diff
@@ -0,0 +1,118 @@
+{
+   "$schema": "http://json-schema.org/schema#",
+   "properties": {
+     "id": {
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-medallia/source_medallia/schemas/feedback.json:4]: Make sure you follow https://docs.airbyte.com/connector-development/schema-reference

File: airbyte-integrations/connectors/source-medallia/source_medallia/source.py [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+
+        data = read_full_refresh(field_id)
+
+        nodes = {}
+
+        for node in data:
+            nodes[node["id"]] = node
+
+        metadata_columns = list(filter(lambda x: x.startswith(tuple(['a_', 'u_', 'e_', 'k_'])), nodes))
+        question_columns = list(filter(lambda x: x.startswith('q_'), nodes))
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-medallia/source_medallia/source.py:238]: Can you create constant tuple like METADATA_PREFIX_COLUMN and QUESTION_PREFIX_COLUMN?

File: airbyte-integrations/connectors/source-medallia/source_medallia/source.py [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+        question_data.field().id()
+        question_data.field().name()
+
+        question_data.__fragment__(extend_field_data)
+
+        feedback.page_info()
+        feedback.page_info.has_next_page()
+        feedback.page_info.end_cursor()
+
+        return {'query': str(op)}
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-medallia/source_medallia/source.py:219]: Please create a unit test to validate this complex function.

File: airbyte-integrations/connectors/source-medallia/source_medallia/source.py [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+        fields.page_info()
+        fields.page_info.has_next_page()
+        fields.page_info.end_cursor()
+
+        return {'query': str(op)}
+
+
+
+
+
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-medallia/source_medallia/source.py:137]: Only 2 lines.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
Sorry the delay to review the contribution. Some comments and need the output from integration tests. Also it is missing the documentation page with the changelog.

File: airbyte-integrations/connectors/source-medallia/source_medallia/source.py [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+        # Other authenticators are available for API token-based auth and Oauth2.
+
+        fields = self.get_fields(config)
+
+        initialization_params = {"authenticator": initialize_authenticator(config), "url_base": config.get("query-endpoint"),
+                                 "fields": fields}
+
+        return [
+            Fields(**initialization_params),
+            Feedback(**initialization_params)
```
[the remark this answers is not in the document]
Comment (JamesGouldSonos) [reply airbyte-integrations/connectors/source-medallia/source_medallia/source.py:267]: Yes the stream just had the two connectors, Survey responses (Feedback) and the Fields information (details about the survey questions.)

File: airbyte-integrations/connectors/source-sap-odata/source_sap_odata/authenticator.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+import logging
+from typing import Any, Mapping
+from urllib import parse
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-sap-odata/source_sap_odata/authenticator.py]: Please remove this

**Pull request closed without merging**